### PR TITLE
[AzureMonitorExporter] Add support new Messaging semantics - Request/Dependency Telemetry

### PR DESCRIPTION
**Pull request by rajkumar-rangaraj**
# Contributing to the Azure SDK

Please see our [CONTRIBUTING.md](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md) if you are not familiar with contributing to this repository or have questions.

For specific information about pull request etiquette and best practices, see [this section](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#pull-request-etiquette-and-best-practices).

### CODE REVIEW

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/AzMonNewListExtensions.cs [hunk 1]
```diff
@@ -15,28 +16,83 @@ internal static class AzMonNewListExtensions
...
+                var isNonDefaultPort = TryGetNonDefaultPort(requestUrlTagObjects[3]?.ToString(), out string? port);
+                var length = (scheme?.Length ?? 0) + (scheme?.Length > 0 ? Uri.SchemeDelimiter.Length : 0) + host.Length + (port?.Length > 0 ? 1 : 0) + (port?.Length ?? 0) + (path?.Length ?? 1) + (queryString?.Length > 0 ? 1 : 0) + (queryString?.Length ?? 0);
+
+                var urlStringBuilder = new System.Text.StringBuilder(length)
+                    .Append(scheme)
+                    .Append(scheme != null ? Uri.SchemeDelimiter : null)
+                    .Append(host)
+                    .Append(isNonDefaultPort ? $":{port}" : null)
+                    .Append(string.IsNullOrEmpty(path) ? "/" : path)
+                    .Append(queryString != null ? $"?{queryString}" : null);
```
Comment (vishweshbankwar) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/AzMonNewListExtensions.cs:36]: `$"?{queryString}"`
Do we need to add `?` here?

Comment (rajkumar-rangaraj) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/AzMonNewListExtensions.cs:36]: `uri.query` does not include `?` in the spec. Hence added manually here. I re-checked the spec now, in one to the example `?` is  a part of it. We need to drive clarity here.

@TimothyMothra  Did you add `?` to the `uri.query` in AspNetCoreInstrumetation or HttpClientInstrumentation?

Comment (vishweshbankwar) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/AzMonNewListExtensions.cs:36]: I think it should be as per [this](https://www.rfc-editor.org/rfc/rfc3986#section-3.4). would be good to test if ASP.NET Core and ASP.NET adds it

Comment (TimothyMothra) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/AzMonNewListExtensions.cs:36]: I'm not seeing any unit tests that check for the query string.
We should include tests in the Exporter

Comment (rajkumar-rangaraj) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/AzMonNewListExtensions.cs]: Current PR covers that test - `GetNewRequestUrl_ReturnsCorrectUrl`

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/AzMonNewListExtensions.cs [hunk 2]
```diff
@@ -88,4 +147,17 @@ private static bool IsDefaultPort(int port)
...
+    {
+        port = stringport;
+
+        if (string.IsNullOrEmpty(stringport) || stringport == "80" || stringport == "443")
+        {
+            return false;
+        }
+
+        return true;
+    }
```
Comment (TimothyMothra) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/AzMonNewListExtensions.cs:162]: I can't find where this is used.

This doesn't look like a typical TryGet pattern.
Why does this need an out param?
If out is needed, should add the `NotNullWhenTrue` attribute.

Comment (rajkumar-rangaraj) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/AzMonNewListExtensions.cs]: Fixed.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/TraceHelper.cs [hunk 3]
```diff
@@ -36,20 +36,29 @@ internal static List<TelemetryItem> OtelToAzureMonitorTrace(Batch<Activity> batc
...
                     {
                         case TelemetryType.Request:
-                            telemetryItem.Data = new MonitorBase
+                            if (activity.Kind == ActivityKind.Server)
                             {
-                                BaseType = "RequestData",
-                                BaseData = new RequestData(Version, activity, ref activityTagsProcessor)
-                            };
+                                var (requestUrl, operationName) = GetHttpOperationNameAndUrl(activity.DisplayName, activityTagsProcessor.activityType, ref activityTagsProcessor.MappedTags);
+                                telemetryItem.Tags[ContextTagKeys.AiOperationName.ToString()] = operationName;
```
Comment (rajkumar-rangaraj) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/TraceHelper.cs:45]: @TimothyMothra We may need a follow up PR here to check if this requires Truncation. Truncation was not a part of existing implementation.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/TraceHelper.cs [hunk 4]
```diff
@@ -36,20 +36,29 @@ internal static List<TelemetryItem> OtelToAzureMonitorTrace(Batch<Activity> batc
...
+                                telemetryItem.Tags[ContextTagKeys.AiLocationIp.ToString()] = TraceHelper.GetLocationIp(ref activityTagsProcessor.MappedTags);
```
Comment (rajkumar-rangaraj) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/TraceHelper.cs:46]: @vishweshbankwar I think for new HTTP semantics, we need to add `client.address`. We could follow up on a new PR.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/TraceHelper.cs [hunk 5]
```diff
@@ -195,6 +210,42 @@ internal static string GetNewSchemaOperationName(Activity activity, string? url,
             return activity.DisplayName;
         }
 
+        internal static (string? RequestUrl, string? OperationName) GetHttpOperationNameAndUrl(string activityDisplayName, OperationType operationType, ref AzMonList httpMappedTags)
```
Comment (rajkumar-rangaraj) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/TraceHelper.cs:213]: The `TelemetryItem` and `RequestData` classes were modified to address an issue where the URL formation was being parsed twice. To resolve this new method was introduced that reads the URL and operation name once for each span / telemetry. This approach eliminates redundant parsing and improves the efficiency of the code for ActivityKind.Server for http spans.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/tests/Azure.Monitor.OpenTelemetry.Exporter.Tests/AzMonNewListExtensionsTests.cs [hunk 6]
```diff
@@ -11,13 +12,11 @@ namespace Azure.Monitor.OpenTelemetry.Exporter.Tests
     public class AzMonNewListExtensionsTests
     {
         [Theory]
-        [InlineData("http", "example.com", "8080", "/search", "q=OpenTelemetry", "http://example.com:8080/search?q=OpenTelemetry")]
-        [InlineData(null, "example.com", "8080", "/search", "q=OpenTelemetry", "example.com:8080/search?q=OpenTelemetry")]
-        [InlineData("http", null, "8080", "/search", "q=OpenTelemetry", null)]
-        [InlineData("http", "example.com", null, "/search", "q=OpenTelemetry", "http://example.com/search?q=OpenTelemetry")]
-        [InlineData("http", "example.com", "8080", null, "q=OpenTelemetry", "http://example.com:8080/?q=OpenTelemetry")]
+        [InlineData("http", "example.com", "8080", "/search", "?q=OpenTelemetry", "http://example.com:8080/search?q=OpenTelemetry")]
```
Comment (rajkumar-rangaraj) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/tests/Azure.Monitor.OpenTelemetry.Exporter.Tests/AzMonNewListExtensionsTests.cs:15]: Removed few inline data here as the `server.address`, `url.scheme` is mandate here.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/tests/Azure.Monitor.OpenTelemetry.Exporter.Tests/TraceHelperNewTests.cs [hunk 7]
```diff
@@ -90,24 +90,26 @@ public void GetNewOperationName_WithNullHttpMethod_ReturnsActivityDisplayName(st
...
+            using var tracerProvider = Sdk.CreateTracerProviderBuilder().AddSource(nameof(GetNewOperationName_WithNullHttpRoute_ReturnsUrlWithVerb)).Build();
+            using var activitySource = new ActivitySource(nameof(GetNewOperationName_WithNullHttpRoute_ReturnsUrlWithVerb));
             using var activity = activitySource.StartActivity(
                 ActivityName,
                 ActivityKind.Server);
             activity?.Stop();
 
             var tagObjects = AzMonList.Initialize();
             AzMonList.Add(ref tagObjects, new KeyValuePair<string, object?>(SemanticConventions.AttributeHttpRequestMethod, "POST"));
+            AzMonList.Add(ref tagObjects, new KeyValuePair<string, object?>(SemanticConventions.AttributeServerAddress, "example.com"));
```
Comment (rajkumar-rangaraj) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/tests/Azure.Monitor.OpenTelemetry.Exporter.Tests/TraceHelperNewTests.cs:105]: `AttributeServerAddress` is a required field, we will not return activity display name.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/tests/Azure.Monitor.OpenTelemetry.Exporter.Tests/CommonTestFramework/TelemetryItemValidationHelper.cs [hunk 8]
```diff
@@ -143,14 +143,6 @@ internal static class TelemetryItemValidationHelper
 
             var expectedTagsCount = 5;
 
-            if (activityKind == ActivityKind.Server)
```
Comment (rajkumar-rangaraj) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/tests/Azure.Monitor.OpenTelemetry.Exporter.Tests/CommonTestFramework/TelemetryItemValidationHelper.cs:146]: This needs to be validated only for HTTP scenario, but this is a manual trace without HTTP semantics hence removed this check.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/TraceHelper.cs (hunk 5 above)
Comment (vishweshbankwar) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/TraceHelper.cs:213]: I think this logic also needs to change. We no longer need to rely on url in new schema as we will have the path available. Simple `http.method + path` concat will be faster/cheaper than `Uri.TryCreate()`.

Can be done as a follow up PR.

Comment (rajkumar-rangaraj) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/TraceHelper.cs:213]: Yep, thought about it and felt better to be followed in a new PR.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/TraceHelper.cs (hunk 4 above)
Comment (vishweshbankwar) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/TraceHelper.cs:52]: Also [UserAgent](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/http.md#:~:text=Recommended-,user_agent.original,-string)

Comment (rajkumar-rangaraj) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/TraceHelper.cs:52]: @TimothyMothra Is there a plan to add `client.address` to `AspNetCoreInstrumentation` library?

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/tests/Azure.Monitor.OpenTelemetry.Exporter.Tests/CommonTestFramework/TelemetryItemValidationHelper.cs (hunk 8 above)
Comment (vishweshbankwar) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/tests/Azure.Monitor.OpenTelemetry.Exporter.Tests/CommonTestFramework/TelemetryItemValidationHelper.cs:146]: I think operation name should be there for consumer spans too. Will check on it.

Comment (rajkumar-rangaraj) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/tests/Azure.Monitor.OpenTelemetry.Exporter.Tests/CommonTestFramework/TelemetryItemValidationHelper.cs:146]: The logic for extracting the operation name is currently based on HTTP semantic attributes, which may not be suitable for custom spans. You're right, it's important to evaluate whether this logic is necessary for custom spans and determine the appropriate value for them.

**Review by vishweshbankwar — APPROVED** [review]

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/TraceHelper.cs (hunk 4 above)
Comment (TimothyMothra) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/TraceHelper.cs:52]: There is no "plan" at the moment.
When working on the instrumentation library, I only converted the attributes that were already in use. I didn't add any new attributes.

The [spec](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.21.0/specification/trace/semantic_conventions/http.md#http-server-semantic-conventions) describes `client.address` as Recommended, but not Required.
Lets discuss this in standup.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/TraceHelper.cs (hunk 3 above)
Comment (TimothyMothra) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/TraceHelper.cs:51]: `Tags` don't have a max length defined in the swagger.
https://github.com/Azure/azure-rest-api-specs/blob/55fc71748c6f51a049fb77aa162ca797aabfbae6/specification/applicationinsights/data-plane/Monitor.Exporters/preview/v2.1/swagger.json#L872C18-L872C18

Comment (TimothyMothra) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/TraceHelper.cs:51]: I found the length for AIOperationName. It's in an internal repo. I'll get this updated

**Pull request merged**